### PR TITLE
Add AdaBelief optimizer

### PR DESCRIPTION
**Pull request by juntang-zhuang**
# Description

Add AdaBelief in tensorflow_addons, which adapts stepsize according to "belief" in gradient, and achieves good generalization, fast convergence and training stability.

Reference: [AdaBelief optimizer: adapting stepsizes by the belief in observed gradients](https://arxiv.org/pdf/2010.07468.pdf) (Juntang Zhuang et al. NeurIPS 2020).

Results: 77.25 (AdaBelief) 77.1 (SGD)
## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [x] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [x] By running pre-commit hooks
- [ ] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?
By running a ResNet50 model from google/tensorflow_models benchmark

If you're adding a bugfix or new feature please describe the tests that you ran to verify your changes:
*

### CODE REVIEW

File: tensorflow_addons/optimizers/adabelief.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""AdaBelief optimizer."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow.compat.v2 as tf
```
Comment (AakashKumarNain) [inline tensorflow_addons/optimizers/adabelief.py:21]: Why compat?

File: tensorflow_addons/optimizers/adabelief.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""AdaBelief optimizer."""
+
+from __future__ import absolute_import
```
Comment (AakashKumarNain) [inline tensorflow_addons/optimizers/adabelief.py:17]: We don't need these as we are strictly on Py3 now

File: tensorflow_addons/optimizers/adabelief.py (hunk 1 above)
Comment (juntang-zhuang) [reply tensorflow_addons/optimizers/adabelief.py:21]: Sorry it's modified from an old branch from google internal source, I'll update it to match the latest version here.

File: tensorflow_addons/optimizers/adabelief.py (hunk 2 above)
Comment (juntang-zhuang) [reply tensorflow_addons/optimizers/adabelief.py:17]: Thanks for reminding, I'll delete it.

File: tensorflow_addons/optimizers/adabelief.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""AdaBelief optimizer."""
+
+from __future__ import division
```
Comment (AakashKumarNain) [inline tensorflow_addons/optimizers/adabelief.py:17]: You can remove all future imports

File: tensorflow_addons/optimizers/adabelief.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+
+import tensorflow as tf
+from tensorflow_addons.utils.types import FloatTensorLike
+
+from typing import Union, Callable, Dict
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class AdaBelief(tf.keras.optimizers.Optimizer):
+    """Variant of the Adam optimizer that achieves fast convergence,
```
Comment (AakashKumarNain) [inline tensorflow_addons/optimizers/adabelief.py:28]: Single line description in the first line e.g. `Variant of the Adam Optimizer.` Then a blank line and then the long description.

File: tensorflow_addons/optimizers/adabelief.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+        )
+        if self.amsgrad:
+            vhat = self.get_slot(var, "vhat")
+            vhat_t = vhat.assign(tf.maximum(vhat, v_t), use_locking=self._use_locking)
+            v_corr_t = tf.sqrt(vhat_t / (1.0 - beta_2_power))
+        else:
+            vhat_t = None
+            v_corr_t = tf.sqrt(v_t / (1.0 - beta_2_power))
+
+        r_t = tf.sqrt(
```
Comment (AakashKumarNain) [inline tensorflow_addons/optimizers/adabelief.py:223]: Here the readability is completely gone. Can you try to make it more readable, split the conditions?

File: tensorflow_addons/optimizers/adabelief.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+
+        if self.amsgrad:
+            vhat = self.get_slot(var, "vhat")
+            vhat_t = vhat.assign(tf.maximum(vhat, v_t), use_locking=self._use_locking)
+            v_corr_t = tf.sqrt(vhat_t / (1.0 - beta_2_power))
+        else:
+            vhat_t = None
+            v_corr_t = tf.sqrt(v_t / (1.0 - beta_2_power))
+
+        r_t = tf.sqrt(
```
Comment (AakashKumarNain) [inline tensorflow_addons/optimizers/adabelief.py:302]: Ditto

File: tensorflow_addons/optimizers/tests/adabelief_test.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests for AdaBelief optimizer."""
+
+from __future__ import absolute_import
```
Comment (AakashKumarNain) [inline tensorflow_addons/optimizers/tests/adabelief_test.py:17]: Remove future imports

File: tensorflow_addons/optimizers/adabelief.py (hunk 5 above)
Comment (juntang-zhuang) [reply tensorflow_addons/optimizers/adabelief.py]: Yeah, thanks for the review. The ```r_t``` computation is actually forked from ```rectified_adam.py```, and I just moved it within the ```if self.rectify``` condition, the idea is to use recitified learning rate as in RAdam when ```self.rectify=True```

For the computation of ```r_t```, do you prefer to keep the current format so that is matches ```RAdam``` exactly, or I break it into shorter lines?

Comment (AakashKumarNain) [reply tensorflow_addons/optimizers/adabelief.py]: I prefer readability. Let's try to break it in a two lines or so

Comment (juntang-zhuang) [reply tensorflow_addons/optimizers/adabelief.py]: Break it into 3 lines to match the numerator, denominator and sqrt operation as in line 11 of Algo2 in https://arxiv.org/pdf/1908.03265.pdf
Please take a look if that reads better now.

Comment (juntang-zhuang) [reply tensorflow_addons/optimizers/adabelief.py]: PS, the long-lines of ```r_t```  is also in ```rectified_adam``` in tensorflow_addons, https://github.com/tensorflow/addons/blob/337951b2371dfa2bc3d4d5cd778de9f25b6f0789/tensorflow_addons/optimizers/rectified_adam.py#L209
Perhaps we can consider format that part too for readability.

Comment (AakashKumarNain) [reply tensorflow_addons/optimizers/adabelief.py]: Yes we should

**Comment by AakashKumarNain** [comment]
@juntang-zhuang please format the code, and make sure your flake8 and black versions are same as in tf-addons

File: tensorflow_addons/optimizers/adabelief.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+            m_t = self._resource_scatter_add(m, indices, m_scaled_g_values)
+        m_corr_t = m_t / (1.0 - beta_1_power)
+
+        v = self.get_slot(var, "v")
+        m_t_indices = tf.gather(m_t, indices)
+        v_scaled_g_values = (
+            tf.math.square(grad - m_t_indices) * (1 - beta_2_t) + epsilon_t
+        )
+        v_t = v.assign(v * beta_2_t, use_locking=self._use_locking)
+        with tf.control_dependencies([v_t]):
```
Comment (AakashKumarNain) [inline tensorflow_addons/optimizers/adabelief.py:281]: Sorry for missing this earlier but in TF2.x, you don't require to use this. We avoid using this in addons as well. Plase check the [docs](https://www.tensorflow.org/api_docs/python/tf/control_dependencies)

File: tensorflow_addons/optimizers/adabelief.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+                r_t * m_corr_t / (v_corr_t + epsilon_t),
+                m_corr_t,
+            )
+        else:
+            var_t = m_corr_t / (v_corr_t + epsilon_t)
+
+        if self._has_weight_decay:
+            var_t += wd_t * var
+
+        with tf.control_dependencies([var_t]):
```
Comment (AakashKumarNain) [inline tensorflow_addons/optimizers/adabelief.py:308]: Same as above

File: tensorflow_addons/optimizers/adabelief.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+                lr_t + decay_rate * tf.minimum(local_step - warmup_steps, decay_steps),
+            )
+
+        sma_inf = 2.0 / (1.0 - beta_2_t) - 1.0
+        sma_t = sma_inf - 2.0 * local_step * beta_2_power / (1.0 - beta_2_power)
+
+        m = self.get_slot(var, "m")
+        m_scaled_g_values = grad * (1 - beta_1_t)
+        m_t = m.assign(m * beta_1_t, use_locking=self._use_locking)
+        with tf.control_dependencies([m_t]):
```
Comment (juntang-zhuang) [inline tensorflow_addons/optimizers/adabelief.py:271]: Do I need to remove all control_dependencies including here? @AakashKumarNain
